pkg/nanny/timer.go

```diff
@@ -19,41 +21,56 @@ type Timer struct {
 	lock sync.Mutex
 }
 
-// MarshalJSON marshals a nanny.Timer into JSON. Fields name, notifier, next_signal and meta are exported
+// MarshalJSON marshals a nanny.Timer into JSON. Fields name, notifier, next_signal, all_clear and meta are exported
 func (nt *Timer) MarshalJSON() ([]byte, error) {
 	return json.Marshal(&struct {
 		Name       string            `json:"name"`
 		Notifier   string            `json:"notifier"`
 		NextSignal string            `json:"next_signal"`
+		AllClear   bool              `json:"all_clear"`
 		Meta       map[string]string `json:"meta,omitempty"`
 	}{
 		Name:       nt.signal.Name,
 		Notifier:   nt.signal.Notifier.String(),
 		NextSignal: nt.end.Format(time.RFC3339),
+		AllClear:   nt.signal.AllClear,
 		Meta:       nt.signal.Meta,
 	})
 }
 
 func newTimer(s validSignal, nanny *Nanny) *Timer {
 	timer := &Timer{signal: s, nanny: nanny}
 	timer.end = time.Now().Add(timer.signal.NextSignal)
-	timer.timer = time.AfterFunc(timer.signal.NextSignal, timer.onExpire)
+	// If NextSignal is in the past but needed for all-clear notification do not notify user until Timer is reset
+	if timer.signal.NextSignal.Seconds() > 0 {
+		timer.timer = time.AfterFunc(timer.signal.NextSignal, timer.onExpire)
+	} else {
+		timer.timer = time.AfterFunc(math.MaxInt64, timer.onExpire)
+		timer.timer.Stop()
+	}
 	return timer
 }
 
 // Reset updates the nannyTimers signal to reset the timer
 func (nt *Timer) Reset(vs validSignal) {
 	nt.lock.Lock()
+	// Send all-clear notification if requested
+	if nt.signal.AllClear && time.Now().After(nt.end) {
+		nt.lock.Unlock()
```

Comment:
Yeah, this unlock for the `nt.notify` call which then locks again is so ugly :/
I understand why you did it like this, but do you think it could be improved?

Maybe if the AllClear signal and implementation were completely new types, which would just trigger different configured message, that would remove the need for these locks?

Please note I've not touched this codebase for over 2 years so I don't have it in RAM anymore 😄